### ISSUE TITLE
Lint idea: Strict `void` checking

### ISSUE THREAD

**Issue by lrhn** [dart-lang/linter#1830]
I would like to avoid the use of the value of a `void` typed expression at all costs, as well as avoid assigning values to `void` contexts.

To that end, I would want the following restrictions (some may overlap with the `void_checks` lint):
* An expression with static type `void` must only occur where its value is not used.
* The only expression allowed for an expression with context type `void` is `null`.
* The only expression allowed for an expression with context type `FutureOr^N<void>` is `null` or a value of type `Future<X>` where `X` is a assignable to `FutureOr^(N-1)<void>`.
* An expression with static type `FutureOr^N<void>` may only occur where its value is not used, or as the operand of an `is`/`is!` check or an `as` cast with a type which is `Future^K<FutureOr^M<void>>` where K >= 1 and K + M <= N.
* A non-`async` function with declared return type `void` must not contain any `return e;` statements, and must not have a `=>` body.
* A non-`async` function with declared return type different from `void` must not contain any `return;` statements.
* An `async` function with a declared return type `T` where `flatten(T)` is `void`, must not contain any `return e;` statements.
* An `async` function with a declared return type `T` where `flatten(T)` is not `void`, must not contain any `return;` statements.

(In short: `void` is not assignable to any type, including itself, and only `Null` is assignable to `void`).

For the above, an expression's value "is not used" iff:
* It is the expression of an ExpressionStatement.
* It is a non-declaration init-part of a `for` loop: `for (here;..;..) ...`.
* It is any of the expressions of the increment part of a `for` loop: `for (..;..;here, andHere) ...`.
* It is the expression of a parenthesized expression, `(here)`, and the value of the parenthesized expression is not used.
* Or it is either of the branch expressions of a conditional expression, `... ? here : orHere`, and the value of the conditional expression is not used.

In particular, I do not want to allow `void foo() => e;` since it is equivalent to `void foo() { return e; }` which is disallowed.

The restrictions may be a single lint, or it may be multiple independent lints.

**Examples**

```dart
class C {
  void foo() {
    if (test) return bar();
  }
  void bar() { ... }
}
```
This returns the value of `bar()`, even though nothing is intended to be returned.
If a subclass overrides `bar` to return a non-`void` type, which it can, then `foo` now returns an actual value. Some user might be tempted into using that value, relying on behavior that was unintended.

```dart
Future<X> foo(void x) async { ...
   if (x is Future) await x;
   return ...;
}
```
This is just bad code, but is currently allowed by the language. The author may have decided at some point that they didn't care about the argument, but then realized that they should check if it was a future anyway. Not wanting to change the API, they just decided to take advantage of the language allowing `is` checks on `void`.

**Additional context**
The current language behavior was chosen as a trade-off between void-soundness and backwards compatibility. This lint is intended as a new start for new code without any backwards compatibility considerations.

Some might disagree with the `=>` rule, wanting to allow `void foo() => expr;` for any expression `expr`, because of the *brevity*. That may be a reasonable trade-off, but I'd prefer if disallowing it was an option as well, perhaps as a separate lint.

**Comment by bwilkerson** [comment]
@MichaelRFairhurst

**Comment by MichaelRFairhurst** [comment]
Detailed thoughts here!

> An expression with static type void must only occur where its value is not used.

This will break mockito and make it unlikely to be widely used:

```
class C {
  void f() {...}
}

class MockC extends Mock implements C {}

test() {
  var c = MockC();
  c.f();
  verify(c.f());
}
```

@srawlins 

> The only expression allowed for an expression with context type void is null

`void_checks` mostly does this. It will allow `void` to be used in a `void` context for the above reason. It does not check `await (void)` because that is covered by the rule `await_only_futures`. I think we typically try to have lints not overlap as it becomes burdensome to use @pq. `void_checks` will allow `is`/`as` checks, too, since the language does, but I think it would be reasonable to restrict this.

> The only expression allowed for an expression with context type FutureOr^N<void> is null or a value of type Future<X> where X is a assignable to FutureOr^(N-1)<void>.

This would be a great addition to `void_checks`. I think it falls under the original intent of that rule.

> An expression with static type FutureOr^N<void> may only occur where its value is not used, or as the operand of an is/is! check or an as cast with a type which is Future^K<FutureOr^M<void>> where K >= 1 and K + M <= N.

Also strikes me as a great addition to `void_checks`.

> A non-async function with declared return type void must not contain any return e; statements, and must not have a => body.

I would highly encourage a user study around `=>`. It is unlikely in my mind that anywhere near a majority of users would find this a readability benefit, or what exactly this would aim to achieve. I would happily get on board if we had data showing that this improved the language.

I would be OK with this being a lint rule like `strict_expression_returns`, if not some `strict_void` rule or something else that has some of the other ideas here.

However, I remember we had to allow this because of lack of time:

```
f() => e;
```

where `f()` gets an _implicit dynamic_. This one is gross and should be fixed.

`return e` is an interesting one. On the one hand, I'm tempted to point to the mockito example and say that sequencing is useful/necessary. On the other hand, banning `return print()` would not itself break mockito. So maybe that could be doable.

> A non-async function with declared return type different from void must not contain any return; statements.

I thought this was already the case?

> An async function with a declared return type T where flatten(T) is void, must not contain any return e; statements.

this would also be a great supplement to `void_checks`, assuming it is consistent with whether it allows `return (void)` or not.

Other thoughts:

I think these are the five biggest problems with void currently:

```
// 1. backwards compatible void returns
f() => print("foo");
g() => print(f()); // oops!

// 2. awaiting void
h() async {
  await print("oops");
}

// 3. FutureOr<void> usability
i() {
  FutureOr<void> fov = print("foo");
  print(fov);
}

// 4. "voidness preservation" never landed
List<void> voidList;
List<dynamic> dynamicList = voidList;

// 5. void is usable inside any generic
class C<T> {
  f(T t) {
    print(t);
  }
}
```

`void_checks` should (not sure if it does) address 1. `await_only_futures` addresses 2 (and I believe this should be part of the language at some point). This issue does a good job of addressing 3. We've talked about a lint for 4 but never gotten to it.

I've always thought 5 should be explicit ie

```
class C<T extends void> {}
```

this could be accomplished via a package:meta annotation `@allowVoid`, optionally coupled with a lint like `strict_void_generics`.

I will say though, I don't think this issue here does a _perfect_ job of addressing issue 3:

```
f() async {
  Object a(bool b) {
    return b ? 1 : Future.value(1);
  }
  FutureOr<void> Function(bool) b = a;
  FutureOr<void> fov = b(random.nextBool());
  assert(fov == null || fov is Future && (await fov) == null));
}
```

So I would personally prefer a lint that simply suggests you avoid `FutureOr<void>` while the language team decides whether `void` should be a sibling type of `Object`/`dynamic`, or at the bottom of the type hierarchy, or unrelated to the type hierarchy.

And as long as void is a supertype of `Future<T>`, the type `FutureOr<void>` will always be the same as `void`, and that is bound to behave in strange ways. The same is true of `FutureOr<Object>` and `FutureOr<dynamic>` etc. But I would rather solve this with algebraic types than keep it as a union type, because I personally like `void` being part of the hierarchy. I can imagine it must be possible to make the VM smart around `FutureOr<T>` to do the right stuff witha sum type here.

```
abstract class FutureOr<T> {}
class Future<T> extends FutureOr<T> {}
abstract class Immediate<T> extends FutureOr<T> {}
```

**Comment by lrhn** [comment]
>  while the language team decides whether void should be a sibling type of Object/dynamic, or at the bottom of the type hierarchy, or unrelated to the type hierarchy.

I am not aware of any discussion about this. The `void` type is top type, mutual subtype of `dynamic` and `Object` (`Object?` when comes NNBD) and a supertype of `Future<void>`, and there are no plans to change that.

The type `FutureOr<void>` is not *the same* as `void` any more than it's the same as `Object`.
You are *expected* to test whether a `FutureOr<void>` is a `Future<void>` or not. If it is not, you should not use it. If so, you should probably `await` it (after promoting to `Future<void>`).
If it was a plain `void` type, you are not expected to look at the value at all. 

Currently there is nothing preventing you from doing other things to a `FutureOr<void>`, which is why I'd want to have stricter lints.

**Comment by MichaelRFairhurst** [comment]
> > while the language team decides whether void should be a sibling type of Object/dynamic, or at the bottom of the type hierarchy, or unrelated to the type hierarchy.
> I am not aware of any discussion about this. The void type is top type

Oops, I meant to say, whether void should be the _top_ type, leaving `dynamic` and `Object` as subtypes (not mutual subtypes) of `void`. And I had only heard mention of it from I think @eernstg, it may have never been a serious consideration.

> The type FutureOr<void> is not the same as void any more than it's the same as Object.

That is the similarity that I'm referring to. It's a very very strong similarity.

If FutureOr were an algebraic type instead of a union, then we wouldn't have many of these problems. Granted we'd have a new set of problems, especially a scary migration :)

**Comment by eernstg** [comment]
Right, `void` is just one of the top types today, and we never went very far with any other idea. If it had been a proper supertype of all other types then we could have detected the difference dynamically, and that would have allowed us to enforce voidness preservation using standard subtype rules. But we didn't do that.

**Comment by MichaelRFairhurst** [comment]
> Currently there is nothing preventing you from doing other things to a FutureOr<void>, which is why I'd want to have stricter lints.

I definitely agree.

I like the idea of

> An expression with static type FutureOr^N<void> may only occur where its value is not used, or as the operand of an is/is! check or an as cast with a type which is Future^K<FutureOr^M<void>> where K >= 1 and K + M <= N.

and this could be either part of `void_checks` or a new name a la `strict_future_or_void`.

I see how the other rules go on to try to protect any value from going into void except null (this is what makes me think about it not being part of the type hierarchy; if it's a singleton it's the unit type. Perhaps we should consider making void a unit type), but I don't see them being conclusive, so I'm not sure if they're worth the inconvenience (especially on mockito users).

**Comment by eernstg** [comment]
We can't treat `void` as the singleton type that just contains the value `null`, because that would make it a subtype of too many other types (all types today, and all types of the form `T?` with nnbd, plus a couple of extras like `dynamic` and `void` and `Null`). So even if we wish to maintain that the value of an expression of type `void` ought to be the null object, we still need to have special rules if we want to help developers enforce the rule that such values should be discarded. (Or, as mentioned, `void` could be a proper supertype of `Object`/`Object?`.)

**Comment by MichaelRFairhurst** [comment]
Right, ideally we would have void be a singleton type whose value can only be instantiated by the compiler.

Perhaps we could have some static getter `void.value` for when users need it.

That would allow mockito/sequencing to still work:

```dart
verify(m.voidMethod());
```

but it would break this pattern which people use:

```dart
void Function() fn = functionThatReturnsNonVoid;
```

so I would rather `void` be a supertype of `Object?`/`dynamic` than a singleton type.

That said, I get confused by `FutureOr<SubtypeOfFuture>` more than `void` being a supertype of everything, and maybe the typical user is more like the opposite.
